scripts/node_js/save_whiteboard_as_png.ts

```diff
@@ -0,0 +1,123 @@
+  fabric[type].fromObject(whiteboardElement, (deserializedElement: any) => {
+    // When all elements have been added to the canvas, we will make sure each has its proper index.
+    deserializedElements.push(deserializedElement)
+
+    const bound = deserializedElement.getBoundingRect()
+    left = Math.min(left, bound.left)
+    top = Math.min(top, bound.top)
+    right = Math.max(right, bound.left + bound.width)
+    bottom = Math.max(bottom, bound.top + bound.height)
+
+    render()
+  })
+})
+
+const render = _.after(whiteboardElements.length, function() {
+  // At this point we've figured out what the left-most and right-most element is. By subtracting
+  // their X-coordinates we get the desired width of the canvas. The height can be calculated in
+  // a similar way by using the Y-coordinates
+  let width = right - left
+  let height = bottom - top
+
+  // Neither width nor height should exceed 2048px.
+  let scaleFactor = 1
+  if (width > 2048 && width >= height) {
+    scaleFactor = 2048 / width
+  } else if (height > 2048 && height > width) {
+    scaleFactor = 2048 / height
+  }
+  if (scaleFactor < 1) {
+    // If scaling down is required, first change the canvas dimensions.
+    width = width * scaleFactor
+    height = height * scaleFactor
+    // Next, scale and reposition each element against top left corner of the canvas.
+    _.each(deserializedElements, function(element) {
+      element.scaleX = element.scaleX * scaleFactor
+      element.scaleY = element.scaleY * scaleFactor
+      element.left = left + ((element.left - left) * scaleFactor)
+      element.top = top + ((element.top - top) * scaleFactor)
+    })
+  }
+  // Add a bit of padding so elements don't stick to the side
+  width += (2 * WHITEBOARD_PADDING)
+  height += (2 * WHITEBOARD_PADDING)
+
+  // Create a canvas and pan it to the top-left corner
+  const canvas = new fabric.Canvas(null, {
+    backgroundColor: '#fff',
+    width: width,
+    height: height,
+  })
+
+  const pt = new fabric.Point(left - WHITEBOARD_PADDING, top - WHITEBOARD_PADDING)
+  canvas.absolutePan(pt)
+
+  // Don't render each element when it's added, rather render the entire Canvas once all elements
+  // have been added. This is significantly faster
+  canvas.renderOnAddRemove = false
+
+  // Once all elements have been added to the canvas, restore
+  // the layer order and convert to PNG
+  const finishRender = _.after(deserializedElements.length, function() {
+    // Ensure each element is placed at the right index. This can only happen
+    // once all elements have been added to the canvas
+    canvas.getObjects().sort(function(elementA, elementB) {
+      return elementA.index - elementB.index
+    })
+
+    // Render the canvas
+    canvas.renderAll()
+    // Convert the canvas to a PNG file and return the data
+    canvas.createPNGStream().pipe(fs.createWriteStream(`${BASE_DIR}/whiteboard.png`))
```

Comment:
This script does work although many details must be worked out. E.g., where to write the PNG file such that the Squiggy py process knows where to find it.